### DISCUSSION TITLE
ADS1115 settings incorrect in source

### DISCUSSION THREAD

**Discussion by zegerk** [arendst/Tasmota#11460] [Q&A] [answered]
Based on the documentation for the ADS1115 [here](https://www.ti.com/lit/ds/symlink/ads1113.pdf?ts=1616606673853&ref_url=https%253A%252F%252Fwww.google.com%252F) the sample rates per second are 8, 16, 32, 64, 128, 250, 475 and 960

However, the following is stated in the source (xsns_12_ads1115.ino)

```#define ADS1115_REG_CONFIG_DR_MASK      (0x00E0)
#define ADS1115_REG_CONFIG_DR_128SPS    (0x0000)  // 128 samples per second
#define ADS1115_REG_CONFIG_DR_250SPS    (0x0020)  // 250 samples per second
#define ADS1115_REG_CONFIG_DR_490SPS    (0x0040)  // 490 samples per second
#define ADS1115_REG_CONFIG_DR_920SPS    (0x0060)  // 920 samples per second
#define ADS1115_REG_CONFIG_DR_1600SPS   (0x0080)  // 1600 samples per second (default)
#define ADS1115_REG_CONFIG_DR_2400SPS   (0x00A0)  // 2400 samples per second
#define ADS1115_REG_CONFIG_DR_3300SPS   (0x00C0)  // 3300 samples per second
#define ADS1115_REG_CONFIG_DR_6000SPS   (0x00E0)  // 6000 samples per second
```

This does not reflect the actual options - after some trial and error, I got the best measurements using 0x0000, using 0x00E0 actually increased the noise level significantly, which is the actual default setting of the module.

**Comment by barbudor** [comment, marked as answer]
Hi @zegerk 
Indeed you touched an interesting point
The driver is named after ADS1115 as the defines in the code and it is supposed to be compatible with both ADS1115 and ADS1015
It is also stated in the top header that the driver is based on on Adafruit's ADS1015's driver
And actually the sample rates listed in the code matches the ADS1015 not the ADS115
ADS1015 datasheet :
![image](https://user-images.githubusercontent.com/2996042/112376214-76e3ba00-8ce4-11eb-94ec-cba13ef9d621.png)
ADS1115 datasheet:
![image](https://user-images.githubusercontent.com/2996042/112376355-a4c8fe80-8ce4-11eb-8aa7-1d352060841a.png)

A brief overlook of the datasheet seems to point this as the only difference between the both ADS.

Considering the driver is reading the ADC only at 250ms period it may seem strange to have such a high sampling frequency.
And your experience seems to confirm that a lower frequency would improve the SNR.

However If you look carefully, you will also see that the driver is performing 2 conversions per channel, waiting 8 ms between each read.
So if you reduce the sampling rate to 8 per seconds, the ADC will not have the time to perform conversion during the 8 ms.
Minimum sampling rate would need to be 128SPS.

So I don't know what you are getting especially if you are using the 4 channels but it cannot be correct with 8SPS.

But this driver definitively needs some love and refactoring....
